Mainboard_ESP32/components/lora/lora.h

```diff
@@ -0,0 +1,266 @@
+// Copyright 2023 PWr in Space, Krzysztof Gliwiński
+#pragma once
+
+#include <stdint.h>
+#include <string.h>
+
+#include "driver/gpio.h"
+#include "driver/spi_master.h"
+#include "esp_log.h"
+#include "esp_system.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "rom/gpio.h"
+#include "soc/gpio_struct.h"
+
+/*
+ * IRQ masks
+ */
+#define IRQ_TX_DONE_MASK 0x08
+#define IRQ_PAYLOAD_CRC_ERROR_MASK 0x20
+#define IRQ_RX_DONE_MASK 0x40
+
+#define PA_OUTPUT_RFO_PIN 0
+#define PA_OUTPUT_PA_BOOST_PIN 1
+
+#define TIMEOUT_RESET 100
+
+/*
+ * Register definitions
+ */
+#define REG_FIFO 0x00
+#define REG_OP_MODE 0x01
+#define REG_FRF_MSB 0x06
+#define REG_FRF_MID 0x07
+#define REG_FRF_LSB 0x08
+#define REG_PA_CONFIG 0x09
+#define REG_LNA 0x0c
+#define REG_FIFO_ADDR_PTR 0x0d
+#define REG_FIFO_TX_BASE_ADDR 0x0e
+#define REG_FIFO_RX_BASE_ADDR 0x0f
+#define REG_FIFO_RX_CURRENT_ADDR 0x10
+#define REG_IRQ_FLAGS 0x12
+#define REG_RX_NB_BYTES 0x13
+#define REG_PKT_SNR_VALUE 0x19
+#define REG_PKT_RSSI_VALUE 0x1a
+#define REG_MODEM_CONFIG_1 0x1d
+#define REG_MODEM_CONFIG_2 0x1e
+#define REG_PREAMBLE_MSB 0x20
+#define REG_PREAMBLE_LSB 0x21
+#define REG_PAYLOAD_LENGTH 0x22
+#define REG_MODEM_CONFIG_3 0x26
+#define REG_RSSI_WIDEBAND 0x2c
+#define REG_DETECTION_OPTIMIZE 0x31
+#define REG_DETECTION_THRESHOLD 0x37
+#define REG_SYNC_WORD 0x39
+#define REG_DIO_MAPPING_1 0x40
+#define REG_VERSION 0x42
+
+/*
+ * Transceiver modes
+ */
+#define MODE_LONG_RANGE_MODE 0x80
+#define MODE_SLEEP 0x00
+#define MODE_STDBY 0x01
+#define MODE_TX 0x03
+#define MODE_RX_CONTINUOUS 0x05
+#define MODE_RX_SINGLE 0x06
+
+/*
+ * PA configuration
+ */
+#define PA_BOOST 0x80
+
+/*!
+  \brief Lora functions return values enum
+*/
+typedef enum {
+  LORA_OK = 0,
+  LORA_INIT_ERR,
+  LORA_WRITE_ERR,
+  LORA_TRANSMIT_ERR,
+  LORA_RECEIVE_ERR
+} lora_err_t;
+
+typedef enum {
+  LORA_GPIO_MODE_DISABLE = 0,
+  LORA_GPIO_MODE_INPUT,
+  LORA_GPIO_MODE_OUTPUT
+} lora_gpio_mode_t;
+
+typedef bool (*lora_SPI_transmit)(uint8_t _in[2], uint8_t _val[2]);
+typedef void (*lora_delay)(size_t _ms);
+typedef bool (*lora_GPIO_set_level)(uint8_t _gpio_num, uint32_t _level);
+typedef void (*lora_GPIO_pad_select_gpio)(uint8_t _gpio_num);
+typedef bool (*lora_GPIO_set_direction)(uint8_t _gpio_num,
+                                        lora_gpio_mode_t _direction);
```

Comment:
pad_select i set_direction można przenieść na użytkownika bo wykorzystujesz to tylko w inicie więc szkoda trzymać cały czas wskaźniki na te funkcje

Reply:
ok